make check_bindist fails to find reference to `main`

#### Description
Cannot run the bindist example.  Via `git bisect` it appears to have stopped working as expected on https://github.com/RIOT-OS/RIOT/commit/452333cb3664a8b301c6957e0aa52422759792e8, under PR #16945 

#### Steps to reproduce the issue
```
cd examples/bindist
make all
make bindist
cd bindist
RIOTBASE=../../../ make check_bindist
```

#### Expected results
All applications compile to a .bin file, and it reports that the bin files match.

#### Actual results
The `make check_bindist` call fails, with an undefined reference to main

#### Versions
```
Operating System Environment
----------------------------
         Operating System: "Ubuntu" "20.04.3 LTS (Focal Fossa)"
                   Kernel: Linux 5.11.0-36-generic x86_64 x86_64
             System shell: /usr/bin/dash (probably dash)
             make's shell: /usr/bin/dash (probably dash)

Installed compiler toolchains
-----------------------------
               native gcc: gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
        arm-none-eabi-gcc: arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 9-2019-q4-major) 9.2.1 20191025 (release) [ARM/arm-9-branch revision 277599]
                  avr-gcc: missing
         mips-mti-elf-gcc: missing
           msp430-elf-gcc: missing
       riscv-none-elf-gcc: missing
  riscv64-unknown-elf-gcc: missing
     riscv-none-embed-gcc: missing
     xtensa-esp32-elf-gcc: missing
   xtensa-esp8266-elf-gcc: missing
                    clang: missing

Installed compiler libs
-----------------------
     arm-none-eabi-newlib: "3.1.0"
      mips-mti-elf-newlib: missing
        msp430-elf-newlib: missing
    riscv-none-elf-newlib: missing
riscv64-unknown-elf-newlib: missing
  riscv-none-embed-newlib: missing
  xtensa-esp32-elf-newlib: missing
xtensa-esp8266-elf-newlib: missing
                 avr-libc: missing (missing)

Installed development tools
---------------------------
                   ccache: missing
                    cmake: cmake version 3.16.3
                 cppcheck: missing
                  doxygen: 1.8.17
                      git: git version 2.25.1
                     make: GNU Make 4.2.1
                  openocd: Open On-Chip Debugger 0.11.0+dev-00347-g3f1c15d2a-dirty (2021-08-24-11:12)
                   python: Python 3.8.10
                  python2: missing
                  python3: Python 3.8.10
                   flake8: error: /usr/bin/python3: No module named flake8
               coccinelle: missing

```

Perhaps a `make check_bindist` should be run in the CI or nightly build to ensure it is kept operational, or at least to have a working example developers can refer to, beyond just the README.

--- a/Makefile.base
+++ b/Makefile.base
@@ -114,7 +114,11 @@ $(BINDIR)/$(MODULE)/:
 	$(Q)mkdir -p $@
 
 OLD_OBJECTS = $(wildcard $(BINDIR)/$(MODULE)/*.o)
-OBJECTS_TO_REMOVE = $(filter-out $(OBJ),$(OLD_OBJECTS))
+
+# do not clean objects from bindist modules
+ifeq (,$(filter $(MODULE),$(BIN_USEMODULE)))
+  OBJECTS_TO_REMOVE = $(filter-out $(OBJ),$(OLD_OBJECTS))
+endif
 
 $(MODULE).module compile-commands $(OBJ): | $(BINDIR)/$(MODULE)/
 

--- a/makefiles/vars.inc.mk
+++ b/makefiles/vars.inc.mk
@@ -21,6 +21,7 @@ export CXXINCLUDES           # The extra include paths for c++, set by the vario
 export NATIVEINCLUDES        # The native include paths, set by the various native Makefile.include files.
 
 export USEMODULE             # Sys Module dependencies of the application. Set in the application's Makefile.
+export BIN_USEMODULE         # Modules specific to bindist (see bindist.ink.mk). Set in the application's Makefile.
 export USEPKG                # Pkg dependencies (third party modules) of the application. Set in the application's Makefile.
 export DISABLE_MODULE        # Used in the application's Makefile to suppress DEFAULT_MODULEs.
 # APPDEPS                    # Files / Makefile targets that need to be created before the application can be build. Set in the application's Makefile.

